### PR TITLE
Fix windows iree+tensorflow build and verify tests.

### PR DESCRIPTION
**Pull request by stellaraccident**
* Switches us to MSVC for the supported windows compiler.
* Uses a couple of magic flags that keep the TF kernels from taking a lot of time to compile and inlining the world to bloat the binary size.
* The whole shebang builds for me now in a few minutes and produces a ~76MB compiler binary. While large, this is manageable for now.
* We still have a fragile relationship with which TF kernels are compiled in and what we can subsequently handle (relying on constant folding to paper over op support issues). We'll need to triage these one by one.
* There are a couple of bugs that seem MSVC/Windows specific -- will need to track down.

### CODE REVIEW

File: .bazelrc [hunk 1]
```diff
@@ -12,38 +12,51 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+###############################################################################
+# Common flags that apply to all configurations.
+# Use sparingly because they cannot be overriden.
```
Comment (GMNGeoffrey) [inline .bazelrc:17]: This shouldn't be true in general. We set `--compilation_mode=opt`, but that doesn't prevent `--compilation_mode=fastbuild` from working.

I guess the issue is that *repeated* flags can't be removed, only overridden, which doesn't work for a copt that the compiler totally doesn't know about.

File: .bazelrc [hunk 2]
```diff
@@ -12,38 +12,51 @@
...
+# Prevent invalid caching if input files are modified during a build.
+build --experimental_guard_against_concurrent_changes
+# Disable visibility checks (works around some private deps in TensorFlow that
+# are being unbundled soon anyway).
+build --nocheck_visibility
+build --compilation_mode=opt
+# Used in TensorFlow, so we have to enable it here as well.
+common --experimental_repo_remote_exec
+# Actually printing output on errors is... a useful default
+test --test_output=errors
```
Comment (GMNGeoffrey) [inline .bazelrc:28]: I usually run with this, but I'm not sure we should be changing the defaults here. This is a development workflow preference thing and seems orthogonal to this change. If I'm running 1000 tests I don't necessary want to see the errors inline. They'll usually be off the screen anyway after the final test results.

File: .bazelrc [hunk 3]
```diff
@@ -12,38 +12,51 @@
...
+build:generic_clang --copt=-Wno-invalid-offsetof
+build:generic_clang --copt=-Wno-unused-function
 
 # C++14 standard version is required.
-build --cxxopt=-std=c++14 --host_cxxopt=-std=c++14
+build:generic_clang --cxxopt=-std=c++14 --host_cxxopt=-std=c++14
 
 # Default to an optimized build with asserts.
 # This is a good compromise between runtime and debugability.
 # Override via: "-c dbg" or --compilation_mode=dbg
```
Comment (GMNGeoffrey) [inline .bazelrc:50]: The "Override" comment doesn't make sense here anymore. It's referring to `--compilation_mode=opt`

File: .bazelrc [hunk 4]
```diff
@@ -159,36 +161,33 @@ build:windows --copt=/DWIN32_LEAN_AND_MEAN --host_copt=/DWIN32_LEAN_AND_MEAN
...
 # "graphics device interface" (GDI) for certain GUI sample projects.
 build:windows --per_file_copt=tensorflow@-DNOGDI
 
-# Used in TensorFlow, so we have to enable it here as well.
-common --experimental_repo_remote_exec
+# Disables TensorFlow eigen bloat and reduces compile times.
+build:windows --define=override_eigen_strong_inline=true
+
+# Another TensorFlow flag from their config script.
+build:windows --define with_default_optimizations=true
```
Comment (GMNGeoffrey) [inline .bazelrc:183]: Seems like these would apply for clang too?

File: .bazelrc [hunk 5]
```diff
@@ -159,36 +161,33 @@ build:windows --copt=/DWIN32_LEAN_AND_MEAN --host_copt=/DWIN32_LEAN_AND_MEAN
...
 
 ###############################################################################
 
+# The user.bazelrc file is not checked in but available for local mods.
+# Always keep this at the end of the file so that user flags override.
+try-import %workspace%/local.bazelrc
+
```
Comment (GMNGeoffrey) [inline .bazelrc:190]: Comment doesn't match the filename. Not sure what the intent is here. Do you want to rename this file?

File: configure_bazel.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
+# Copyright 2020 Google LLC
```
Comment (GMNGeoffrey) [inline configure_bazel.py:1]: Can this go in build_tools?

Comment (benvanik) [reply configure_bazel.py:1]: yeah build_tools/bazel/ seems nice

File: .bazelrc (hunk 1 above)
Comment (stellaraccident) [reply .bazelrc:17]: Rephrased.

File: .bazelrc (hunk 2 above)
Comment (stellaraccident) [reply .bazelrc:28]: I feel moderately strongly about this. I've seen enough engineers be shocked and awed to find out that such a thing exists to treat it as a poor default. If people would like it the other way, they can disable it :)

File: .bazelrc (hunk 3 above)
Comment (stellaraccident) [reply .bazelrc:50]: Remove.

File: .bazelrc (hunk 4 above)
Comment (stellaraccident) [reply .bazelrc:183]: Maybe? I haven't tested them and have done builds on linux that have done fine without them. I'll leave them for windows until the next time I am building on another platform.

File: configure_bazel.py (hunk 6 above)
Comment (stellaraccident) [reply configure_bazel.py:1]: I feel moderately strongly (and there is plenty of precedent) that levers needed to meaningfully do anything with the project belong in the root. Having a configure script in the root is a pretty strong signal that some action is expected to be taken prior to just jumping in.

**Comment by stellaraccident** [comment]
PTAL

**Comment by antiagainst** [comment]
Yay! No need to set the env beforehand is nice. I kinda always forget that. (Mostly using CMake though.)

File: .bazelrc [hunk 7]
```diff
@@ -118,37 +131,26 @@ build:rs --bes_results_url="https://source.cloud.google.com/results/invocations/
 build:rs --project_id=iree-oss
 
 ###############################################################################
+# Windows specific flags for building with VC.
```
Comment (ScottTodd) [inline .bazelrc:134]: This is awesome :D

I tested locally (threadripper 3970x so unfair comparison with other CPUs).

My builds before this were ~1h, with nearly all of the time spent compiling a long tail of TF kernels on a few threads. The build would run in parallel efficiently for ~10 minutes then be basically idle working on ~8 kernel files for 50 minutes.

Test steps (using Colab as the target since it pulls in most of the relevant deps):
```
bazel clean --expunge
python3 configure_bazel.py
python3 .\colab\start_colab_kernel.py
```

user.bazelrc (note that I use separate caches and output_user_root for separate repositories/branches to try to keep as much state cached per branch as possible)
```
startup --output_user_root=D:/dev/cache/bazeluserroot
build --disk_cache=D:/dev/cache/bazelcache --experimental_guard_against_concurrent_changes
build:debug --compilation_mode=dbg --copt=/O2 --per_file_copt=iree@/Od --strip=never
build --define=iree_tensorflow=true
```

Timing:

```
8:58 start
  300s "Fetching @aws;"
9:03 finished fetching, build running (128 actions in parallel)
9:04 821/2656
9:05 3747/4793 (working on tf targets)
9:06 4005/4793 (tf kernels, core/util/batch_util.cc)
9:07 4885/5342 (starting llvm/lib/CodeGen/...)
9:08 5275/5726 (more LLVM)
9:09 5893/6256 (LLVM, MLIR, IREE)
9:10 6693/6842 (IREE, XLA, MLIR, linking IREE python bindings)
finished 9:10
```

File: docs/GetStarted/getting_started_linux_bazel.md [hunk 8]
```diff
@@ -48,6 +48,7 @@ Clone the repository and initialize its submodules:
 $ git clone https://github.com/google/iree.git
 $ cd iree
 $ git submodule update --init
+$ python3 configure_bazel.py
```
Comment (ScottTodd) [inline docs/GetStarted/getting_started_linux_bazel.md:51]: nit: the text above this "Clone the repository and initialize its submodules" could be expanded

File: docs/GetStarted/getting_started_windows_bazel.md [hunk 9]
```diff
@@ -66,6 +61,7 @@ clone the repository and initialize its submodules:
 > git clone https://github.com/google/iree.git
 > cd iree
 > git submodule update --init
+> python configure_bazel.py
```
Comment (ScottTodd) [inline docs/GetStarted/getting_started_windows_bazel.md:64]: nit: the text above this "clone the repository and initialize its submodules" could be expanded

File: .bazelrc (hunk 7 above)
Comment (ScottTodd) [reply .bazelrc:134]: (super scientific :P looks like https://docs.bazel.build/versions/master/skylark/performance.html#profile-and-analyze-profile would give more concrete data)

File: configure_bazel.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  if platform.system() == "Windows":
+    # Sure. Good idea, bazel.
+    return p.replace("\\", "/")
+
+
+def write_platform(bazelrc):
+  platform_config = "generic_clang"
+  if platform.system() == "Windows":
+    platform_config = "windows"
+  print("build --config={}".format(platform_config), file=bazelrc)
```
[the remark this answers is not in the document]
Comment (stellaraccident) [reply configure_bazel.py:31]: yeah

File: docs/GetStarted/getting_started_linux_bazel.md [hunk 11]
```diff
@@ -42,12 +42,12 @@ export CXX=clang++
...
 ### Clone
 
-Clone the repository and initialize its submodules:
+Clone the repository, initialize its submodules and configure:
 
 ```shell
 $ git clone https://github.com/google/iree.git
 $ cd iree
-$ git submodule update --init
+$ git submodule update --initss
```
Comment (ScottTodd) [inline docs/GetStarted/getting_started_linux_bazel.md:50]: Is there a typo here now? (this is pending merge, can fixup after)

Comment (stellaraccident) [reply docs/GetStarted/getting_started_linux_bazel.md:50]: ugh. for some reason vscode stopped accepting ctrl+s and I was looking everywhere for where the extra 's''s went. sorry :(
